kaiax/builder/interface.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2025 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package builder
+
+import (
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/kaiax"
+)
+
+type TxGenerator func(nonce uint64) (*types.Transaction, error)
+
+//go:generate mockgen -destination=./mock/module.go -package=mock github.com/kaiachain/kaia/kaiax/builder BuilderModule
+type BuilderModule interface {
+	kaiax.BaseModule
+	kaiax.JsonRpcModule
+
+	// IncorporateBundleTx does the followings:
+	IncorporateBundleTx(txs []*types.Transaction, bundles []*Bundle) []interface{}
+
+	// Arrayify flattens txs in heap into an array
+	Arrayify(heap *types.TransactionsByPriceAndNonce) []*types.Transaction
+
+	// IsConflict checks if the new bundles conflict with the previous bundles
+	IsConflict(prevBundles []*Bundle, newBundles []*Bundle) bool
+}
+
+type TxBundlingModule interface {
+	// The function finds transactions to be bundled.
+	// New transactions can be injected.
+	// returned bundles must not have conflict with `prevBundles`.
+	// `txs` and `prevBundles` is read-only; it is only to check if there's conflict between new bundles.
+	ExtractTxBundles(txs []*types.Transaction, prevBundles []*Bundle) []*Bundle
+}
```

Comment:
Since this may be implemented by multiple modules, shouldn't we move it to kaiax/interface.go?
I also plan to define an interface such as `TxBundlingModuleHost` and have `worker` implement it.
https://github.com/kaiachain/kaia/blob/a127f8bfac6e764768b06aa74564bf53ea8e2498/kaiax/interface.go#L86-L89

Reply:
However, it seems possible to have a host interface within a module, like the staking module. Is the difference between these two versatility?
https://github.com/kaiachain/kaia/blob/a127f8bfac6e764768b06aa74564bf53ea8e2498/kaiax/staking/interface.go#L51-L53

Reply:
IMO we can view `kaiax/interface.go` as a fallback, which is used only if we don't have anywhere else to put `*ModuleHost`. This can prevent `kaiax/interface.go` from growing too large.

Reply:
Rule of thumb is to define AModule and AModuleHost at the same file. In that perspective, TxBundlingModuleHost should be in kaiax/builder/interface.go